net processing: Reduce resource usage for inbound block-relay-only connections

jnewbery opened the pull request:
block-relay-only connections are additional outbound connections that bitcoind makes since v0.19. They participate in block relay, but do not propagate transactions or addresses. They were introduced in #15759.

When creating an outbound block-relay-only connection, since we know that we're never going to announce transactions over that connection, we can save on memory usage by not a `TxRelay` data structure for that connection. When receiving an inbound connection, we don't know whether the connection was opened by the peer as block-relay-only or not, and therefore we always construct a `TxRelay` data structure for inbound connections.

However, it is possible to tell whether an inbound connection will ever request that we start announcing transactions to it. The `fRelay` field in the `version` message may be set to `0` to indicate that the peer does not wish to receive transaction announcements. The peer may later request that we start announcing transactions to it by sending a `filterload` or `filterclear` message, **but only if we have offered `NODE_BLOOM` services to that peer**. `NODE_BLOOM` services are disabled by default, and it has been recommended for some time that users not enable `NODE_BLOOM` services on public connections, for privacy and anti-DoS reasons.

Therefore, if we have not offered `NODE_BLOOM` to the peer _and_ it has set `fRelay` to `0`, then we know that it will never request transaction announcements, and that we can save resources by not initializing the `TxRelay` data structure.

jnewbery committed:
[net processing] Don't initialize TxRelay for non-tx-relay peers.

Delay initializing the TxRelay data structure for a peer until we receive
a version message from that peer. At that point we'll know whether it
will ever relay transactions. We only initialize the m_tx_relay
data structure if:

- this isn't an outbound block-relay-only connection; AND
- fRelay=true OR we're offering NODE_BLOOM to this peer
  (NODE_BLOOM means that the peer may turn on tx relay later)

jnewbery committed:
[net processing] Comment all TxRelay members

This fully comments all the TxRelay members. The only significant change
is to the comment for m_relay_txs. Previously the comment stated that
one of the purposes of the field was that "We don't relay tx invs before
receiving the peer's version message". However, even without the
m_relay_txs flag, we would not send transactions to the peer before
receiving the `version` message, since SendMessages() returns
immediately if fSuccessfullyConnected is not set to true, which only
happens once a `version` and `verack` message have been received.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2643,10 +2642,17 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        // We don't initialize the m_tx_relay data structure if:
+        // - this is an outbound block-relay-only connection; or
+        // - fRelay=false and we're not offering NODE_BLOOM to this peer
+        //   (NODE_BLOOM means that the peer may turn on tx relay later)
```
in 238ce0b02b1f7ac78c8c0466cef33628c6d49359:

If it is turned on later, we could keep it a nullptr until then and only initialize if and when we receive a `filter*` message?

jnewbery replied:
Perhaps, but I think it's better to allocate resources for the worst-case scenario up front, rather than try to reduce resource usage and later find out that you need to allocate a new structure.

naumenkogs replied:
>rather than try to reduce resource usage and later find out that you need to allocate a new structure

What exactly is difficult about this?

MarcoFalke replied:
> What exactly is difficult about this?

I think I agree with John that there is little benefit (or even a risk) when giving remote peers the option to force you to allocate MBs of memory by sending a single msg, instead of allocating the memory on startup/creation of the connection.

vasild replied:
Hmm, allocating upfront is like giving a remote peer the option and he always exercises it immediately after connecting, no?

MarcoFalke replied:
Yes, and I think this is preferable because there are less (late) surprises. Neither on the live network, nor when developing. For example, `massif` will display the real usage from the start. Or, imagine someone under-estimating the resource usage and thus increasing the maxconnections count, then running into OOM later on.

vasild replied:
The counter scenario is somebody running into OOM because MBs of memory were allocated needlessly (allocated but never used because the peer(s) never sent `filter*`).

MarcoFalke replied:
I think if you want to reduce memory usage, you should disable bip37, not enable it with the hope that no one (or only a small fraction of your incoming peers) will use it.

maflcko reviewed (commented):
Concept ACK

jnewbery committed:
[net processing] Add m_tx_relay_mutex to protect m_tx_relay ptr

naumenkogs commented:
Concept ACK

jnewbery commented:
Rebased on #21160

vasild commented:
Some rough estimates on how much this will save:

```
(gdb) p sizeof(CNode::TxRelay)
$1 = 152
```

Assuming 50 incoming connections and 20% of them being block-only, that makes 10 peers will benefit from this. 152 bytes each, so 1520 bytes will be saved.

jnewbery commented:
> Some rough estimates on how much this will save...

@vasild This seems completely wrong. `TxRelay` contains a CRollingBloomFilter parametrized by `{50000, 0.000001}`. According to this code comment:

https://github.com/bitcoin/bitcoin/blob/419afa93419e6840f78cb94b4a39d826eb10e139/src/bloom.h#L94-L98

the storage requirement for just that rolling bloom filter is ~1.8 * 50,000 * 6 = 540kB

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -227,28 +227,45 @@ struct Peer {
+        /** Set of transaction ids we still have to announce (txid for
+         *  non-wtxid-relay peers, wtxid for wtxid-relay peers). We use the
+         *  mempool to sort transactions in dependency order before relay, so
+         *  this does not have to be sorted. */
```
At first I was thinking we could leave a TODO here for switching to unordered_set, but when I'm looking how it's used, it seems we can just leave it like this, right?

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -227,28 +227,45 @@ struct Peer {
         std::chrono::microseconds m_next_inv_send_time{0};
 
-        /** Minimum fee rate with which to filter inv's to this node */
+        /** Minimum fee rate with which to filter transaction announcements to this node */
```
nit: this could reference BIP133

naumenkogs replied:
This also could use BIP133, or perhaps be flipped with the next variable referencing it :)

vasild commented:
Well, I said "rough", but did not mean _that_ rough. Somehow I wrongly assumed that none of the constructors allocate extra memory from the heap :disappointed:. Now I checked that `CRollingBloomFilter{50000, 0.000001}` creates and fills a vector of 67'396 `uint64_t` elements, or 539'168 bytes. For 10 peers that would save ~5MB.

I wonder if it would make sense to delay the creation of `CRollingBloomFilter::data` until it is actually needed (e.g. create it in the first insert, rather than in the constructor, would be changes isolated solely inside `CRollingBloomFilter`). `CRollingBloomFilter` is also used here:

* `BanMan::m_discouraged`, would benefit from lazy creation because we may never ban anybody, but is just one filter per program instance.
* `Peer::m_addr_known`, looks like if created then it will be certainly used, one per peer.
* `PeerManagerImpl::m_recent_rejects` and `PeerManagerImpl::m_recent_confirmed_transactions`, likely to be used, one per program instance.
* `CNodeState::m_recently_announced_invs`, would benefit like `CNode::TxRelay::filterInventoryKnown`, one per peer.

Maybe only `CNodeState::m_recently_announced_invs` is worth considering?

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -227,28 +227,45 @@ struct Peer {
+         *  announcements. It is *not* a p2p protocol violation for the peer to
+         *  send us transactions with a lower fee rate than this. See BIP133. */
         CAmount m_fee_filter_sent{0};
+        /** The next time after which we will send a `feefilter` message to this peer. */
```
I think this comment is not very useful currently w.r.t. why we would update feefilter at all, and why it has to be scheduled.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2658,8 +2657,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        // - this is an outbound block-relay-only connection; or
+        // - fRelay=false and we're not offering NODE_BLOOM to this peer
+        //   (NODE_BLOOM means that the peer may turn on tx relay later)
+        const bool disable_tx_relay = pfrom.IsBlockOnlyConn() ||
```
Inverted this could be:
`const bool enable_tx_relay = !pfrom.IsBlockOnlyConn() && (fRelay || (pfrom.GetLocalServices() & NODE_BLOOM))`.

Don't you think it's better?

MarcoFalke commented:
> @vasild  e.g. create it in the first insert, rather than in the constructor

See discussion at https://github.com/bitcoin/bitcoin/pull/22778#discussion_r693960652

naumenkogs commented:
ACK 7ef58ab5ca3e0652442eaf56c5c91ce192df324a

jnewbery commented:
rebased

jnewbery commented:
rebased

mzumsande commented on src/net_processing.cpp:
```diff
@@ -2710,8 +2709,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        // - fRelay=true or we're offering NODE_BLOOM to this peer
+        //   (NODE_BLOOM means that the peer may turn on tx relay later)
+        const bool enable_tx_relay = !pfrom.IsBlockOnlyConn() &&
+                                      (fRelay || (pfrom.GetLocalServices() & NODE_BLOOM));
```
I think that means peers to which we'd connect in `-blocksonly` mode, also wouldn't initialize `m_tx_relay` anymore. 
But `-blocksonly` mode until now allows to submit and relay transactions via RPC (`sendrawtransaction`), even if it that is obviously bad for privacy. 
With this change, attempting to send a tx via RPC in `-blocksonly` mode might lead to all of our peers disconnecting us because they wouldn't have `m_tx_relay` enabled anymore.

jnewbery replied:
Good catch. I've updated the `inv` and `tx` logic to explicitly test whether the peer is a block-relay-only connect instead of testing the existence of the Peer::TxRelay struct. That means that if we send a transaction to a peer while we're in blocksonly mode, then they will still process it.

MarcoFalke replied:
>  With this change, attempting to send a tx via RPC in -blocksonly mode might lead to all of our peers disconnecting us because they wouldn't have m_tx_relay enabled anymore.

Isn't there a test already that should check for this?

jnewbery replied:
>> With this change, attempting to send a tx via RPC in -blocksonly mode might lead to all of our peers disconnecting us because they wouldn't have m_tx_relay enabled anymore.

> Isn't there a test already that should check for this?

I would guess that:

https://github.com/bitcoin/bitcoin/blob/bdbabc50ba6c87ded97ea2bbacd3605c59cd12d0/test/functional/p2p_blocksonly.py#L43

would catch it. This branch will previously a draft since it was shared as a WIP, and I'm not sure if that test was failing.

mzumsande replied:
I did check that when I wrote my earlier comment: That line didn't fail because the peer we sent the tx to was just a python `P2PInterface()`, which wouldn't disconnect us for sending a tx like a full node would have.

jnewbery replied:
I guess the test could be changed to subclass `P2PInterface` and fail if the node under test sends a tx message.

jnewbery commented:
Rebased now that #21160 has been merged. I haven't addressed @mzumsande's comment https://github.com/bitcoin/bitcoin/pull/22778#discussion_r780726527 yet, so leaving in draft.

jnewbery commented:
Thanks for the review @mzumsande. I've responded to your review comment. Let m know what you think.

I've also marked this as ready for review.

Note the behaviour change for `feefilter` messages. With the current branch, for a peer that indicates that they don't want us to relay messages with fRelay=false in the `version` message, we will no longer send `feefilter` messages. I think that's fine, but it can easily be changed by pulling the `m_fee_filter_sent` and `m_next_send_feefilter` members out of `Peer::TxRelay` and into `Peer`. That makes sense since all other members of `TxRelay` are concerned with relaying transactions _to_ the peer, and those two members are concerned with receiving transactions _from_ the peer, so they don't really belong int `TxRelay`.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2714,10 +2743,21 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            (fRelay || (pfrom.GetLocalServices() & NODE_BLOOM))) {
+            auto* tx_relay = [&peer]() {
+                LOCK(peer->m_tx_relay_mutex);
+                Assume(!peer->m_tx_relay);
```
#21596 :smiling_face_with_tear: 

```
net_processing.cpp:2754:31: error: reading variable 'm_tx_relay' requires holding mutex 'peer->m_tx_relay_mutex' [-Werror,-Wthread-safety-analysis]
                Assume(!peer->m_tx_relay);
                              ^
1 error generated.

jnewbery replied:
Yeah, I'll remove the lambda.

MarcoFalke replied:
I think the issue isn't with the lambda, but with `Assume`/`Assert`.

jnewbery replied:
Thanks. Removed.

MarcoFalke replied:
Just for reference, an alternative to removing would be to temporarily asssign the value to a bool and then `Assume(name_of_bool)`.

MarcoFalke replied:
Another reference: If #24714 is merged, you may be able to re-add the Assume as it was.

jnewbery replied:
I'll leave this for now, and add the `Assume()` back if 24714 is merged. I'm 100% sure that this code path can only get hit once for each peer, so the `Assume()` was only there to catch weird regressions.

jnewbery replied:
> I'll leave this for now, and add the Assume() back if 24714 is merged.

Now done.

jnewbery commented:
Rebased now that #24714 has been merged. I've also added a setter method for Peer.m_tx_relay and made the member variable private.

dergoegge reviewed (commented):
Concept ACK

I think moving `m_fee_filter_sent` and `m_next_send_feefilter` out of `TxRelay` as you mentioned in https://github.com/bitcoin/bitcoin/pull/22778#issuecomment-1082831424 would be good and probably the right change. To me, a PR that only introduced the behavior change for feefilter messages would seem unnecessary, so i am leaning more twoards not introducing the change here. A peer that sends `fRelay=false` could still send us transactions, why should we not let them know what our min fee is?

jnewbery committed:
[net processing] [refactor] Move m_next_send_feefilter and m_fee_filter_sent

Move m_next_send_feefilter and m_fee_filter_sent out of the `TxRelay`
data structure. All of the other members of `TxRelay` are related to
sending transactions _to_ the peer, whereas m_fee_filter_sent and
m_next_send_feefilter are both related to receiving transactions _from_
the peer. A node's tx relay behaviour is not always symmetrical (eg a
blocksonly node will ignore incoming transactions, but may still send
out its own transactions), so it doesn't make sense to group the
feefilter sending data with the TxRelay data in a single structure.

This does not change behaviour, since IsBlockOnlyConn() is always equal
to !peer.m_tx_relay. We still don't send feefilter messages to outbound
block-relay-only peers (tested in p2p_feefilter.py).

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2744,7 +2753,13 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        // - this isn't an outbound block-relay-only connection; and
+        // - fRelay=true or we're offering NODE_BLOOM to this peer
+        //   (NODE_BLOOM means that the peer may turn on tx relay later)
+        if (!pfrom.IsBlockOnlyConn() &&
```
I agree that we could skip it, and also agree that it's out of scope for this PR (which shouldn't change behaviour).

Also note that we only have at most one feeler connection at a time, so this wouldn't result in much resource saving (whereas it's potentially very significant for inbound block-relay-only peers, since there could be many of them).

jnewbery commented:
> I think moving m_fee_filter_sent and m_next_send_feefilter out of TxRelay as you mentioned in https://github.com/bitcoin/bitcoin/pull/22778#issuecomment-1082831424 would be good and probably the right change. To me, a PR that only introduced the behavior change for feefilter messages would seem unnecessary, so i am leaning more twoards not introducing the change here. A peer that sends fRelay=false could still send us transactions, why should we not let them know what our min fee is?

ok, updated this PR to move those fields out of the TxRelay struct. This PR should not change the feefilter behaviour for inbound or outbound block-relay-only peers.

dergoegge reviewed (commented):
Code review ACK ee135a2a67b96024b66b094d28f9ab3ec6d634be

jnewbery commented on src/net_processing.cpp:
```diff
@@ -235,36 +235,64 @@ struct Peer {
-        // a) it allows us to not relay tx invs before receiving the peer's version message
-        // b) the peer may tell us in its version message that we should not relay tx invs
-        //    unless it loads a bloom filter.
+        /** Whether the peer wishes to receive transaction announements.
```
Thanks. Fixed.

dergoegge commented:
Code review ACK 4bebd4c72c9328ec6dc7ef5d8d1dd7b53d73fa22

Only change since i last reviewed was fixing a typo.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -235,32 +235,48 @@ struct Peer {
-        // Last time a "MEMPOOL" request was serviced.
+        /** The last time a BIP35 `mempool` request was serviced. */
         std::atomic<std::chrono::seconds> m_last_mempool_req{0s};
+        /** The next time after which we will send an `inv` message containing
```
Nit: it also determines sending mempool response through `fSendTrickle`
```

// Respond to BIP35 mempool requests
if (fSendTrickle && peer->m_tx_relay->m_send_mempool) {
```

naumenkogs replied:
Arguably, this field could be moved out of `TxRelay` then. Or perhaps we should use another timer for mempool.

jnewbery replied:
> Nit: it also determines (through fSendTrickle) sending mempool response

The response to a mempool message is one or more inv announcement messages for transactions in the mempool. I think the comment is correct.

> Arguably, this field could be moved out of TxRelay then. Or perhaps we should use another timer for mempool.

I think it'd be better to simply remove support for the `mempool` message. I think it's unlikely that many people are using it, and it'd be better to provide this functionality (exposing the contents of the mempool) through rpc rather than a 'private' p2p extension.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -3353,7 +3368,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         // Stop processing the transaction early if
         // 1) We are in blocks only mode and peer has no relay permission
```
nit: add `OR` to the comment

jnewbery replied:
Done

naumenkogs commented:
Code review ACK 4bebd4c72c9328ec6dc7ef5d8d1dd7b53d73fa22

------
There seem to be a mistake in the commit message of the latter commit. It should say "fRelay=true OR we're offering", as opposed to "fRelay=true OR we're not offering". The same comment in the codebase is correct.


```
We only initialize the m_tx_relay
data structure if:

- this isn't an outbound block-relay-only connection; AND
- fRelay=true OR we're not offering NODE_BLOOM to this peer
  (NODE_BLOOM means that the peer may turn on tx relay later)
  ```

jnewbery commented:
> There seem to be a mistake in the commit message of the latter commit. It should say "fRelay=true OR we're offering", as opposed to "fRelay=true OR we're not offering". 

Good catch. Thanks! I've fixed the commit.

Also rebased on latest master.

naumenkogs commented:
reACK c4455b7d86b782b7c010857e04be5a73dd1dbb43

maflcko reviewed (approved):
looking good. review ACK c4455b7d86b782b7c010857e04be5a73dd1dbb43, some doc feedback 🚓

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

looking good. review ACK c4455b7d86b782b7c010857e04be5a73dd1dbb43, some doc feedback 🚓
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhgTQv9EW6RubZe9M3V45udjFTMx1NNg/Chg5MNKvi5Old75iobKOtqJ+4mrgAx
lBNIwRaJaW8JLuf173YQyQqWYcwcGJnbmN7xStR7OSJ7a0Z8LBSBab30yOGHhspy
eAV9vJSxzdgq7flinRA1rJe1CR8zupDqGgMbAWdqsWo+HyQcdiM2HAy5QINRCw3Y
ykENjyjQFytzEBub9GFBrwe95hrJGvQhNO4ewazNvecZtTf0gllq/PM5MUANb+Av
J9nzFORGRBmAb2d/7ARIGGGwC1qA6I4vsk0DaWH8MfbGwIcA+GnbUuZW+aSh2al9
Prk7r1aUY+BczUcb5WqzoS4NRVF7x6MGdrY9FJBoXYrow/SNIPF3iTjuFJ8NSgD2
7v9liK6adx2g8Jz7MO5+UjgrDXXDEAKLvz68mSaVktZghpjWQbKdqo9v/oi6INni
noBAcGDMNmGg+hg9lGmjmRmHvc1z6Kj15qmqDtDP4+Vl82vpbOHFwotynkc6qMAM
nMIbay5Q
=JQJR
-----END PGP SIGNATURE-----
```


</details>

jnewbery commented on src/net_processing.cpp:
```diff
@@ -237,32 +237,48 @@ struct Peer {
-    /** The feerate in the most recent BIP133 `feefilter` message sent to the peer. */
+    /** The feerate in the most recent BIP133 `feefilter` message sent to the peer.
+     *  It is *not* a p2p protocol violation for the peer to send us
+     *  transactions with a lower fee rate than this. See BIP133. */
```
Agree. This was a rebase oopsie from moving commits around. Fixed.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -237,32 +237,48 @@ struct Peer {
     CAmount m_fee_filter_sent{0};
-    /** Timestamp to send the next BIP133 `feefilter` message sent to the peer. */
+    /** Timestamp after which we will send the next BIP133 `feefilter` message
+      * to the peer. */
```
as above

jnewbery commented on src/net_processing.cpp:
```diff
@@ -237,32 +237,48 @@ struct Peer {
     std::chrono::microseconds m_next_send_feefilter{0};
 
     struct TxRelay {
+        /** Protects m_bloom_filter and the m_relay_txs flag. */
```
removed

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2056,7 +2096,7 @@ void PeerManagerImpl::ProcessGetData(CNode& pfrom, Peer& peer, const std::atomic
 
-        if (peer.m_tx_relay == nullptr) {
+        if (tx_relay == nullptr) {
             // Ignore GETDATA requests for transactions from blocks-only peers.
```
Thanks. Updated.

jnewbery commented:
Thanks for the review @MarcoFalke. All review comments have been addressed.

MarcoFalke commented:
You did the rebase on the wrong machine. The typo from https://github.com/bitcoin/bitcoin/pull/22778#issuecomment-1124711567 is back again.




review ACK edf29446658ee9105b6451573687dc74a92604b9 💎

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK edf29446658ee9105b6451573687dc74a92604b9 💎
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiqcQv+NNN0jFiVXfsQikTdVkZEyp8Re02fU5bsk1AYZlB45koiyFQehpyC87wS
e9vTQpGGy3lu3JoYNjuIeJSC07MPTy8anqunZ2tHfe0xz9XCksqqml4DiFWHhPIr
tLnRxuamaBUy3VzDZN1hulLsG2JI/NB4TuxTEkZOjesiPqiAep79ugKCvm7828At
EYyS/NVO7uQq2x0li/1KFtzRbC9wG/32AkNYMa7je+dlMkrE8InHcLqhCpQp2c89
8dhmUfKMYcxock8yTaQQH6iqEH9vxSBYccLHd9rOpDCrxYcQW39u/EPom4Esqvna
O48dF3aV+lS/ztEB0qH5TtE9p9AdTPZaUw2lHwTzcZ4VgKgD2IyJzemA6a4Wnao/
ymBpWk9RJoToBpKUelqMrkrLhXneIDAtAO28aiAAi0SuNsBxIbLLgweYUemom11L
P/78PSLAEe3QvVpubDPULGXR2JMueWkcHYy09/TT4s44J4DnD/bUF4mpy6vz3fI9
ZuJzQdNA
=3z05
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2051,8 +2088,9 @@ void PeerManagerImpl::ProcessGetData(CNode& pfrom, Peer& peer, const std::atomic
-        if (peer.m_tx_relay == nullptr) {
-            // Ignore GETDATA requests for transactions from blocks-only peers.
+        if (tx_relay == nullptr) {
+            // Ignore GETDATA requests for transactions from blocks-only peers and
```
```suggestion
            // Ignore GETDATA requests for transactions from block-relay-only peers and
```

nit in the last commit: Maybe clarify while touching this line

MarcoFalke commented:
review ACK 9db82f1bca0bb51c2180ca4a4ad63ba490e79da4 🖖

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 9db82f1bca0bb51c2180ca4a4ad63ba490e79da4 🖖
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUia8gv/Y52Be4OmX6OF+MVJkvO5yvf1pgfk+5UrMwxwRHD+JPU/j09OSlRh9Iqh
VQfCIISSQgLbmxXuOODYgIPxK3WDkTBSH1zTNZWxfcODcXr539kTsJMq0Uh40TbI
hPIR+BrUJ+NTW1zQOJtMJeUHPsWTuomqbqYJ6iX360WnUE5D/mi/m2rBaOaq97De
XDb059Ww4NgYmQVJJk3xbqiNskVG/sZkW78W0676+Gum+awt0+XK5aJTH/rFHAio
uDmpAPPGMfnxcNnOi3rj0d/nI/lEBYX1+/0hlnhOZI6y2ordKokRe7ePDSok1MgK
piRGIPbmHzptjnVWvBFybTN7k+zlMBmouG498mmwS2HAw/EaNE+WV0gK1JHycnj5
538RWYKG/qaKc8dPHFjvCnV2QXhmfcoADYnPGmvqVqdS40aAT77upwDd3KB3PYm5
0Jm22Y+CbqT1y6NdUIV0YPVKrjylNrls6FMN/RnBcVNna/iWvR+UBI/sF3pMQiz+
i/LQZkTg
=/JFT
-----END PGP SIGNATURE-----
```


</details>

dergoegge commented:
Core review ACK 9db82f1bca0bb51c2180ca4a4ad63ba490e79da4

Only documentation changed since I last reviewed.

LarryRuane commented:
I was just speaking with @glozow, and I'd like to host review club for this PR on June 8, if it's not already merged by then, or maybe even if it is.

naumenkogs commented:
ACK 9db82f1bca0bb51c2180ca4a4ad63ba490e79da4

jnewbery commented:
> I was just speaking with @glozow, and I'd like to host review club for this PR on June 8, if it's not already merged by then, or maybe even if it is.

That's great! I'm very happy to review notes & questions, or have a call to discuss. Feel free to message me.

LarryRuane commented:
Post-merge ACK
I'd appreciate it if a maintainer could add the review club label.
